src/Discord/Internal/Rest/Guild.hs

```diff
@@ -320,6 +329,137 @@ instance FromJSON ListActiveThreads where
     ListActiveThreads <$> o .: "threads"
                       <*> o .: "members"
 
+-- | Audit log object, along with the entries it also contains referenced users, integrations [...] and so on
+data AuditLog = AuditLog
+  { auditLogEntries             :: [AuditLogEntry]
+  , auditLogUsers               :: [User]
+  , auditLogIntegrations        :: [Integration]
+  , auditLogWebhooks            :: [Webhook]
+  , auditlogScheduledEvents     :: [ScheduledEvent]
+  , auditLogThreads             :: [Channel]
+  , auditLogApplicationCommands :: [ApplicationCommand]
+  , auditLogAutoModerationRules :: [AutoModerationRule]
+  } deriving ( Show, Read )
+
+
+instance FromJSON AuditLog where
+  parseJSON = withObject "AuditLog" $ \o ->
+    AuditLog <$> o .: "audit_log_entries"
+             <*> o .: "users"
+             <*> o .: "integrations"
+             <*> o .: "webhooks"
+             <*> o .: "guild_scheduled_events"
+             <*> o .: "threads"
+             <*> o .: "application_commands"
+             <*> o .: "auto_moderation_rules"
+
+-- | An audit log entry object, so to speak the actual event that took place
+data AuditLogEntry = AuditLogEntry
+  { auditLogEntryId         :: AuditLogEntryId
+  , auditLogEntryActionType :: AuditLogEvent
+  , auditLogEntryUserId     :: Maybe UserId
+  , auditLogEntryTargetId   :: Maybe Snowflake
+  , auditLogEntryChanges    :: Maybe [AuditLogChange]
+  , auditLogEntryOptions    :: Maybe AuditLogEntryOptions
+  , auditLogEntryReasons    :: Maybe String
+  } deriving ( Show, Read)
+
+instance FromJSON AuditLogEntry where
+  parseJSON = withObject "AuditLogEntry" $ \o ->
+    AuditLogEntry <$> o .:  "id"
+                  <*> o .:  "action_type"
+                  <*> o .:? "user_id"
+                  <*> o .:? "target_id"
+                  <*> o .:? "changes"
+                  <*> o .:? "options"
+                  <*> o .:? "reasons"
+
+newtype AuditLogEvent = MkAuditLogEvent Int
+  deriving ( Show, Read )
+
+instance FromJSON AuditLogEvent where
+  parseJSON = fmap MkAuditLogEvent . parseJSON
+
+-- | A change object, it is only partially implemented (see 'AuditLogChangeValue')
+data AuditLogChange = AuditLogChange
+  { auditLogChangeKey      :: String
+  , auditLogChangeNewValue :: Maybe AuditLogChangeValue
+  , auditLogChangeOldValue :: Maybe AuditLogChangeValue
+  } deriving ( Show, Read )
+
+
+instance FromJSON AuditLogChange where
+  parseJSON = withObject "AuditLogChange" $ \o ->
+    AuditLogChange <$> o .:  "key"
+                   <*> o .:? "new_value"
+                   <*> o .:? "old_value"
+
+
+-- | This is a temporary solution, i really hope.
+--   Discord did NOT think this through, read more at <https://discord.com/developers/docs/resources/audit-log#audit-log-change-object>
+data AuditLogChangeValue = AuditLogChangeValue String
+  deriving ( Show, Read )
+
+instance FromJSON AuditLogChangeValue where
+  parseJSON o = return $ AuditLogChangeValue (show o)
```

Comment:
ah I see what the issue is with the change object. Because just about anything can be marked as changed or whatever, just about any object can be included here. I think it would be for the best that the `Value` is stored instead of auditlogchangevalue, since the user can at least use aeson operations on it themselves if they wish.

Reply:
That is definitely a better solution!
Sad that discord put no thought behind this though...
I'll change that shortly